drgn/helpers/linux/rss.py

```diff
@@ -0,0 +1,148 @@
+
+from drgn import Object, Program, TypeKind
+from drgn.helpers.common.prog import takes_program_or_default
+from drgn.helpers.linux.list import list_for_each_entry
+from drgn.helpers.linux.percpu import percpu_counter_sum
+
+__all__ = ("get_task_rss",)
+
+
+class TaskRss(NamedTuple):
+    """
+    Represent's a task's resident set size in pages. See task_rss().
+    """
+
+    rss_file: int
+    rss_anon: int
+    rss_shmem: int
+    swapents: int
+
+    @property
+    def total(self) -> int:
+        return self.rss_file + self.rss_anon + self.rss_shmem
+
+
+def for_each_task_in_group(
+    task: Object, include_self: bool = False
+) -> Iterable[Object]:
+    """
+    Iterate over all tasks in the thread group
+
+    Or, in the more common userspace terms, iterate over all threads of a
+    process.
+
+    :param task: a task whose group to iterate over
+    :param include_self: should ``task`` itself be returned
+    :returns: an iterable of every thread in the thread group
+    """
+    if include_self:
+        yield task
+    if hasattr(task, "thread_group"):
+        yield from list_for_each_entry(
+            "struct task_struct",
+            task.thread_group.address_of_(),
+            "thread_group",
+        )
+    else:
+        # Since commit 8e1f385104ac0 ("kill task_struct->thread_group") from
+        # 6.7, the thread_group list is gone, replaced by a list inside the
+        # task.signal struct. This has an explicit list_head (unlike the
+        # thread_group which just linked each task together with no explicit
+        # head node).
+        for other in list_for_each_entry(
+            "struct task_struct",
+            task.signal.thread_head.address_of_(),
+            "thread_node",
+        ):
+            # We've already yielded "task" (or not, depending on the caller's
+            # preference) so skip it here.
+            if other != task:
+                yield other
+
+
+@takes_program_or_default
+def get_task_rss(
+    prog: Program, task: Object, cache: Optional[Dict[int, TaskRss]]
+) -> TaskRss:
+    """
+    Return the task's resident set size (RSS) in pages
+
+    The task's RSS is the number of pages which are currently resident in
+    memory. The RSS values can be broken down into anonymous pages (not bound to
+    any file), file pages (those associated with memory mapped files), and
+    shared memory pages (those which aren't associated with on-disk files, but
+    belonging to shared memory mappings). This function returns a tuple
+    containing each category, but the common behavior is to use the "total"
+    value which sums them up.
+
+    :param task: ``struct task_struct *`` for which to compute RSS
+    :param cache: if provided, we can use this to cache the mapping of
+      "mm_struct" to RSS. This helps avoid re-computing the RSS value for
+      processes with many threads, but note that it could result in out of date
+      values on a live system.
```

Comment:
I would simply drop the `cache` parameter. drgn itself doesn't usually cache things for you (`identify_address()` is a notable exception). drgn-tools can wrap this with its own cache if needed.